# Wikipedia Notable Life Expectancies

# [Notebook 4 of 4: Data Cleaning](https://github.com/teresahanak/wikipedia-life-expectancy/blob/main/wp_life_expect_data_clean3_thanak_2022_06_23.ipynb)

## Context

The


## Objective

The

### Data Dictionary

- Feature: Description

## Importing Necessary Libraries

In [1]:
# To structure code automatically
%load_ext nb_black

# To import/export sqlite databases
import sqlite3 as sql

# To save/open python objects in pickle file
import pickle

# To help with reading, cleaning, and manipulating data
import pandas as pd
import numpy as np
import re

# To define maximum number of columns to be displayed in a dataframe
pd.set_option("display.max_columns", None)
# To define the maximum number of rows to be displayed in a dataframe
pd.set_option("display.max_rows", 200)

# To supress warnings
# import warnings

# warnings.filterwarnings("ignore")

# To set some visualization attributes
pd.set_option("max_colwidth", 150)

# To play auditory cue when cell has executed, has warning, or has error and set chime theme
import chime

chime.theme("zelda")

<IPython.core.display.Javascript object>

## Data Overview

### Reading, Sampling, and Checking Data Shape

In [2]:
# Reading the dataset
conn = sql.connect("wp_life_expect_clean2.db")
data = pd.read_sql("SELECT * FROM wp_life_expect_clean2", conn)

# Making a working copy
df = data.copy()

# Checking the shape
print(f"There are {df.shape[0]} rows and {df.shape[1]} columns.")

# Checking first 2 rows of the data
df.head(2)

There are 132652 rows and 21 columns.


day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month info_parenth info_1           info_2  \
0  1994  January         None   None   British dancer   
1  1994  January         None   None  Irish economist   

                         info_3        info_4 info_5 info_6 info_7 info_8  \
0  ballet designer and director          None   None   None   None   None   
1                        writer  and academic   None   None   None   None   

  info_9 info_10 info_11   age cause_of_death  
0   None    None    None  86.0           None  
1   None    None    None  68.0           None

<IPython.core.display.Javascript object>

In [3]:
# Checking last 2 rows of the data
df.tail(2)

day           name  \
132650   9  Oleg Moliboga   
132651   9       Zou Jing   

                                                                         info  \
132650          , 69, Russian volleyball player, Olympic champion  and coach.   
132651  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  \
132650        https://en.wikipedia.org/wiki/Oleg_Moliboga              2   
132651  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3   

        year month info_parenth info_1                     info_2  \
132650  2022  June       (1980)   None  Russian volleyball player   
132651  2022  June         None   None           Chinese engineer   

                                              info_3 info_4 info_5 info_6  \
132650                   Olympic champion  and coach   None   None   None   
132651  member of the Chinese Academy of Engineering   None   None   None   

       info_7 info_8 info_9 info_10 info_11   age cause_of_death  
132650   None   None   None    None    None  69.0           None  
132651   None   None   None    None    None  86.0           None

<IPython.core.display.Javascript object>

In [4]:
# Checking a sample of the data
df.sample(5)

day              name  \
93104   30  Chingiz Sadykhov   
40141   12      Kent Douglas   
98667    7     Eija Krogerus   
47787   30  Jenny Wood-Allen   
132024   9            Qin Yi   

                                                                                                       info  \
93104                                                                            , 88, Azerbaijani pianist.   
40141                                                             , 73, Canadian ice hockey player , cancer   
98667                                                                                 , 86, Finnish bowler.   
47787   , 99, Scottish athlete and politician, world record holder for the oldest female marathon finisher.   
132024                                                                             , 100, Chinese actress .   

                                                  link num_references  year  \
93104   https://en.wikipedia.org/wiki/Chingiz_Sadykhov              1  2017   
40141       https://en.wikipedia.org/wiki/Kent_Douglas             10  2009   
98667      https://en.wikipedia.org/wiki/Eija_Krogerus              1  2018   
47787   https://en.wikipedia.org/wiki/Jenny_Wood-Allen              5  2010   
132024            https://en.wikipedia.org/wiki/Qin_Yi             21  2022   

           month           info_parenth info_1  \
93104   December                   None   None   
40141      April  (Toronto Maple Leafs)   None   
98667    October                   None   None   
47787   December                   None   None   
132024       May                 (, , )   None   

                                 info_2  \
93104               Azerbaijani pianist   
40141        Canadian ice hockey player   
98667                    Finnish bowler   
47787   Scottish athlete and politician   
132024                  Chinese actress   

                                                             info_3 info_4  \
93104                                                          None   None   
40141                                                        cancer   None   
98667                                                          None   None   
47787   world record holder for the oldest female marathon finisher   None   
132024                                                         None   None   

       info_5 info_6 info_7 info_8 info_9 info_10 info_11    age  \
93104    None   None   None   None   None    None    None   88.0   
40141    None   None   None   None   None    None    None   73.0   
98667    None   None   None   None   None    None    None   86.0   
47787    None   None   None   None   None    None    None   99.0   
132024   None   None   None   None   None    None    None  100.0   

       cause_of_death  
93104            None  
40141            None  
98667            None  
47787            None  
132024           None

<IPython.core.display.Javascript object>

### Checking Data Types, Duplicates, and Null Values

In [5]:
# Checking data types and null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132652 entries, 0 to 132651
Data columns (total 21 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   day             132652 non-null  object 
 1   name            132652 non-null  object 
 2   info            132652 non-null  object 
 3   link            132652 non-null  object 
 4   num_references  132652 non-null  object 
 5   year            132652 non-null  int64  
 6   month           132652 non-null  object 
 7   info_parenth    49830 non-null   object 
 8   info_1          35 non-null      object 
 9   info_2          132604 non-null  object 
 10  info_3          62571 non-null   object 
 11  info_4          12605 non-null   object 
 12  info_5          1497 non-null    object 
 13  info_6          216 non-null     object 
 14  info_7          31 non-null      object 
 15  info_8          6 non-null       object 
 16  info_9          1 non-null       object 
 17  info_10   

<IPython.core.display.Javascript object>

#### Loading `nation_country_dict` from Pickle File to Dictionary `nation_map`

In [6]:
# Load the nation_country_dict
with open("nation_country_dict.pkl", "rb") as f:
    nation_map = pickle.load(f)

<IPython.core.display.Javascript object>

## Extracting Nationality Continued
Here is the approach we will take:
- The plan will be to save the country name, in lieu of nationality, in new `place_1` and `place_2` columns as it is standardized for the various associated nationality values.
- First, we will update the keys and values in `nation_map` by replacing hyphens with a single space.
- Then we will remove "-born" from the column we are searching, as well as replace "-" and "/" each with single spaces.  In this step, we can also remove leading and trailing periods and whitespace.
- We will proceed to search the numbered `info` columns in order checking as follows:
    1. if column value starts with a value in the dictionary:
        - save country to `place_1` and remove value from searched column.
    2. if `place_1` value has been found:
        - if updated column value starts with a value in the dictionary:
            - save country to `place_2` and remove value from searched column.
    3. Repeat steps 1 and 2 but comparing with country (dictionary keys)
    4. Check unique values for column starting with capital letters.
- It is tempting to shorten the process by simply searching for nationality and country values within the column value, but as there are entries containing first and second nationality, we have to proceed from left to right to extract these values optimally.

#### Removing "-" and "." from `nation_map`

In [7]:
# Removing hyphens from nation_map
nation_map = {
    key.replace("-", ""): value.replace("-", " ") for key, value in nation_map.items()
}

# Removing periods from nation_map
nation_map = {
    key.replace(".", ""): value.replace(".", " ") for key, value in nation_map.items()
}

<IPython.core.display.Javascript object>

#### Removing or Replacing Extra Characters in Numbered `info` Columns

In [8]:
%%time

# List of columns to treat
cols_lst = [
    "info_1",
    "info_2",
    "info_3",
    "info_4",
    "info_5",
    "info_6",
    "info_7",
    "info_8",
    "info_9",
    "info_10",
    "info_11",
]

# Dictionary of keys to find and values to replace keys
replace_dict = {'-born': '', '–born': '', '-': ' ', '–': ' ', '/': ' ', '.': ' '}

# For loop to find and replace characters in replace_dict in columns in cols_list
# and strip any leading or trailing periods or whitespace
for column in cols_lst:
    for key, value in replace_dict.items():
        for index in df[column].notna().index:
            item = df.loc[index, column]
            if item:
                df.loc[index, column] = item.replace(key, value).strip()
                
# Chime notification when cell successfully executes
chime.success()

CPU times: total: 1min 43s
Wall time: 1min 43s


<IPython.core.display.Javascript object>

#### Checking `info_1` for `place_1`

In [9]:
# Column to check
column = "info_1"

# Extract to column
extract_to = "place_1"

# Dataframe to check
dataframe = df[(df[column].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if item.startswith(nationality):
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(nationality, "").strip()
            )

# Check a sample of rows
df[df[extract_to].notna()].sample(2)

day            name                                 info  \
32981  28  David Turnbull  . 92, American materials scientist.   
32318   6   Harry Webster   , 89. British automotive engineer.   

                                                                     link  \
32981  https://en.wikipedia.org/wiki/David_Turnbull_(materials_scientist)   
32318                         https://en.wikipedia.org/wiki/Harry_Webster   

      num_references  year     month info_parenth               info_1 info_2  \
32981              8  2007     April         None  materials scientist   None   
32318              1  2007  February         None  automotive engineer   None   

      info_3 info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  \
32981   None   None   None   None   None   None   None    None    None  92.0   
32318   None   None   None   None   None   None   None    None    None  89.0   

      cause_of_death                                               place_1  
32981           None                              United States of America  
32318           None  United Kingdom of Great Britain and Northern Ireland

<IPython.core.display.Javascript object>

#### Observations:
- `info_1` provides us a small sample on which to test code.
- We successfully extracted those `place_1` values, now we will do the same on the treated rows for `place_2`.

#### Checking `info_1` for `place_2`

In [10]:
# Column to check
column = "info_1"

# Extract to column
extract_to = "place_2"

# Dataframe to check
dataframe = df[(df["place_1"].notna()) & (df[column].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if item.startswith(nationality):
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(nationality, "").strip()
            )

# Check a sample of rows
df.sample(2)

day        name                                                 info  \
48649    4  Ed Manning  , 68, American basketball player , heart condition.   
113407  14   Juan Uder  , 93, Argentine basketball player, world champion .   

                                            link num_references  year  month  \
48649   https://en.wikipedia.org/wiki/Ed_Manning              9  2011  March   
113407   https://en.wikipedia.org/wiki/Juan_Uder              3  2020   July   

                                             info_parenth info_1  \
48649   (Baltimore Bullets) and coach (San Antonio Spurs)   None   
113407                                             (1950)   None   

                             info_2           info_3 info_4 info_5 info_6  \
48649    American basketball player  heart condition   None   None   None   
113407  Argentine basketball player   world champion   None   None   None   

       info_7 info_8 info_9 info_10 info_11   age cause_of_death place_1  
48649    None   None   None    None    None  68.0           None     NaN  
113407   None   None   None    None    None  93.0           None     NaN

<IPython.core.display.Javascript object>

#### Observations:
- Here we can see that the new column place_2 has not yet been added as there were not any matching values.
- Let us confirm by checking the remaining unique values in info_1.

#### Checking Remaining Unique Values in `info_1`

In [11]:
# Checking unique values
df["info_1"].unique()

array([None, 'politician', 'Olympic sprinter', 'gridiron football player',
       'writer', 'businessman', 'social psychologist', 'King of Nepal',
       'Maori leader', 'artist', 'English sports journalist',
       'Jules Engel', 'early', 'aka', 'Jr', 'professional wrestler',
       'automotive engineer', 'materials scientist', 'weightlifter',
       'common chimpanzee', '', 'Olympic athlete', 'actor',
       'Olympic gymnast', 'broadcaster and writer', 'Olympic swimmer',
       'Olympic boxer', 'Olympic wrestler', 'Olympic sailor',
       'basketball player', 'college basketball coach',
       'choral conductor', 'Tree of the Year'], dtype=object)

<IPython.core.display.Javascript object>

#### Obsservations:
- Neither "English" nor "Maori" are keys in the current dictionary.
- Maori is an ethnicity within the country of New Zealand, so for now, we will add it as a key our dictionary with the country value of New Zealand.  If we have matching first and second countries, we can later remove the first value.
- We will also add the key "English" with the country value 'United Kingdom of Great Britain and Northern Ireland'.
- Then, we can rerun the above code for `place_1` and `place_2`.
- The country value of "Nepal" is also present. We will hold off on extracting country names until we have first exhausted matching nationalities, as the Wikipedia field called for nationalities.

#### Updating `nation_map`

In [12]:
# Adding key: country pairs to nation_map
nation_map["English"] = nation_map["British"]
nation_map["Maori"] = nation_map["New Zealand"]

<IPython.core.display.Javascript object>

#### Re-checking `info_1` for `place_1`

In [13]:
# Column to check
column = "info_1"

# Extract to column
extract_to = "place_1"

# Dataframe to check
dataframe = df[(df[extract_to].isna()) & (df[column].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if item.startswith(nationality):
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(nationality, "").strip()
            )

# Check a sample of rows
df[df[extract_to].notna()].sample(2)

day          name                           info  \
103467  26  Edmund Seger  , 82 German Olympic wrestler.   
61945    1   Basil Soper        , British actor, 74–75.   

                                              link num_references  year month  \
103467  https://en.wikipedia.org/wiki/Edmund_Seger              2  2019   May   
61945    https://en.wikipedia.org/wiki/Basil_Soper              0  2013  June   

       info_parenth            info_1 info_2 info_3 info_4 info_5 info_6  \
103467         None  Olympic wrestler   None   None   None   None   None   
61945          None             actor   None   None   None   None   None   

       info_7 info_8 info_9 info_10 info_11   age cause_of_death  \
103467   None   None   None    None    None  82.0           None   
61945    None   None   None    None    None  74.5           None   

                                                     place_1  
103467                                               Germany  
61945   United Kingdom of Great Britain and Northern Ireland

<IPython.core.display.Javascript object>

#### Re-checking `info_1` for `place_2`

In [14]:
# Column to check
column = "info_1"

# Extract to column
extract_to = "place_2"

# Dataframe to check
dataframe = df[(df["place_1"].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if item.startswith(nationality):
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(nationality, "").strip()
            )

# Checking rows
df[df["place_2"].notna()]

day                name                             info  \
19580  20  Dame Miraka Szászy  , 80. New Zealand Maori leader.   

                                                 link num_references  year  \
19580  https://en.wikipedia.org/wiki/Mira_Sz%C3%A1szy             21  2001   

          month info_parenth  info_1 info_2 info_3 info_4 info_5 info_6  \
19580  December         None  leader   None   None   None   None   None   

      info_7 info_8 info_9 info_10 info_11   age cause_of_death      place_1  \
19580   None   None   None    None    None  80.0           None  New Zealand   

           place_2  
19580  New Zealand

<IPython.core.display.Javascript object>

#### Observations:
- Our code appears to be finding the matching values and assigning the corresponding country to the correct nation column.
- We see "New Zealand" added to both nation columns here, which was expected as both New Zealand and Maori are in the description
- As an aside, we will need to check our final values where `place_1` is "American" and `place_2` is "Indian" as our code will indicate United States and India, which may or may not be correct. 
- Now we can proceed to doing the same extraction on `info_2`.

#### Checking `info_2` for `place_1`

In [15]:
%%time

# Column to check
column = "info_2"

# Extract to column
extract_to = "place_1"

# Dataframe to check
dataframe = df[(df[extract_to].isna()) & (df[column].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if item.startswith(nationality):
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(nationality, "").strip()
            )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 3min 50s
Wall time: 3min 50s


<IPython.core.display.Javascript object>

In [16]:
# Check a sample of rows
df[df[extract_to].notna()].sample(2)

day                name  \
75819   5     Eshel Ben-Jacob   
38270   2  Kenneth P. Johnson   

                                                     info  \
75819                            , 63, Israeli physicist.   
38270  , 74, American newspaper editor , heart infection.   

                                                   link num_references  year  \
75819     https://en.wikipedia.org/wiki/Eshel_Ben-Jacob              5  2015   
38270  https://en.wikipedia.org/wiki/Kenneth_P._Johnson              6  2008   

          month info_parenth info_1            info_2           info_3 info_4  \
75819      June         None   None         physicist             None   None   
38270  November           ()   None  newspaper editor  heart infection   None   

      info_5 info_6 info_7 info_8 info_9 info_10 info_11   age cause_of_death  \
75819   None   None   None   None   None    None    None  63.0           None   
38270   None   None   None   None   None    None    None  74.0           None   

                        place_1 place_2  
75819                    Israel     NaN  
38270  United States of America     NaN

<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for place_1

In [17]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')

There are 2394 remaining missing values for place_1.



<IPython.core.display.Javascript object>

#### Observations:
- We have captured the `place_1` value for the vast majority of entries.
- Before checking for other variations on nationality usage, let us check to see if the value starts with the country name.

#### Re-checking `info_2` for `place_1` Using Country Value

In [18]:
%%time

# Column to check
column = "info_2"

# Extract to column
extract_to = "place_1"

# Dataframe to check
dataframe = df[(df[extract_to].isna()) & (df[column].notna())]

# For loop to extract nation data to place column
for country in nation_map.values():
    for index in dataframe.index:
        item = df.loc[index, column]
        if item.startswith(country):
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(country, "").strip()
            )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 4.34 s
Wall time: 4.35 s


<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for `place_1`

In [19]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')

There are 2275 remaining missing values for place_1.



<IPython.core.display.Javascript object>

#### Observations:
- We captured over 100 more values for place_1 by directly checking the country name.
- Now, we will examine the remaining capitalized first words in info_1 and append our dictionary as needed.

#### Examining Unique Values of First Word in `info_1` if Upper Case

In [20]:
# Column to check
column = "info_2"

# Dataframe to check
dataframe = df[(df["place_1"].isna()) & (df[column].notna())]

# Checking set of first words in info_2 where place_1 is missing
first_words = set([item.split()[0] for item in dataframe[column] if item[0].isupper()])
print(f"There are {len(first_words)} unique values for first word in info_1.\n")
first_words

There are 302 unique values for first word in info_1.



{'AIDS',
 'ANC',
 'Abkhaz',
 'Abkhazian',
 'Aboriginal',
 'Actress',
 'African',
 'Afrikaans',
 'Afrikaner',
 'Afro',
 'Air',
 'Alfa',
 'All',
 'Alyawarre',
 'Amateur',
 'America',
 "America's",
 'Amrican',
 'Anglican',
 'Anglo',
 'Anguillan',
 'Arabic',
 'Archbishop',
 'Archdeacon',
 'Argentinian',
 'Aruba',
 'Aruban',
 'Assamese',
 'Associate',
 'Assyrian',
 'Athletics',
 'Aussie',
 'Austro',
 'Avarian',
 'Azorean',
 'BBC',
 'Baltic',
 'Basque',
 'Bavarian',
 'Benedictine',
 'Bermudan',
 'Bermudian',
 'Bessarabian',
 'Bletchley',
 'Bodo',
 'Bosnia',
 'Breton',
 'Brigadier',
 "Britain's",
 'Britsih',
 'California',
 'Californian',
 'Calypso',
 'Cantonese',
 'Caribbean',
 'Catalan',
 'Catholic',
 'Caymanian',
 'Ceylon',
 'Ceylonese',
 'Chagossian',
 'Chairman',
 'Chechen',
 'Cherokee',
 'Chief',
 'Chilian',
 'China',
 'Chiricahua',
 'Chuvash',
 'Circassian',
 'Civil',
 'Columbian',
 'Commandant',
 'Composer',
 'Computer',
 'Congo',
 'Congoleze',
 'Congresswoman;',
 'Cook',
 'Cornish',


<IPython.core.display.Javascript object>

#### Observations:
- We can see there are some remaining variations on how nationality was entered that are not yet in `nation_country_dict`.
- Let us add those now, then do another iteration for searching `info_2`.
- Descriptions will be assigned to their geographical physical region where sovereign state is remote or nationality description is broad.

#### Hard-coding Additional Variations on Nationality

In [21]:
# Hard-coding remaining unique nationality/location descriptors
nation_map["ANC"] = nation_map["South African"]
nation_map["Abkhaz"] = nation_map["Georgian"]
nation_map["Abkhazian"] = nation_map["Georgian"]
nation_map["Aboriginal"] = nation_map["Australian"]
nation_map["African"] = "Africa"
nation_map["Afrikaans"] = nation_map["African"]
nation_map["Afrikaner"] = nation_map["African"]
nation_map["Afro"] = "Africa"
nation_map["Alyawarre"] = nation_map["Australian"]
nation_map["America"] = nation_map["US"]
nation_map["America's"] = nation_map["US"]
nation_map["Amrican"] = nation_map["US"]
nation_map["Anglo"] = "Europe"
nation_map["Anguillan"] = "Central America and the Caribbean"
nation_map["Antigua"] = nation_map["Antiguan"]
nation_map["Arabic"] = "Arab states"
nation_map["Argentinian"] = nation_map["Argentine"]
nation_map["Aruba"] = "Central America and the Caribbean"
nation_map["Aruban"] = nation_map["Aruba"]
nation_map["Assamese"] = nation_map["Indian"]
nation_map["Assyrian"] = "Middle East"
nation_map["Aussie"] = nation_map["Australian"]
nation_map["Australia"] = nation_map["Australian"]
nation_map["Australia's"] = nation_map["Australian"]
nation_map["Austria"] = nation_map["Austrian"]
nation_map["Austro"] = nation_map["Austrian"]
nation_map["Avarian"] = nation_map["Russian"]
nation_map["Azerbaijan"] = nation_map["Azerbaijani"]
nation_map["Azorean"] = nation_map["Portuguese"]
nation_map["Baltic"] = "Europe"
nation_map["Bangladesh"] = nation_map["Bangladeshi"]
nation_map["Barbados"] = "Central America and the Caribbean"
nation_map["Basque"] = "Europe"
nation_map["Bavarian"] = nation_map["German"]
nation_map["Belarus"] = nation_map["Belarusian"]
nation_map["Belarussian"] = nation_map["Belarusian"]
nation_map["Belgium"] = nation_map["Belgian"]
nation_map["Bermudan"] = "Central America and the Caribbean"
nation_map["Bermudian"] = "Central America and the Caribbean"
nation_map["Bessarabian"] = "Europe"
nation_map["Bletchley"] = nation_map["British"]
nation_map["Bodo"] = nation_map["Norwegian"]
nation_map["Bosnia"] = nation_map["Bosnian"]
nation_map["Breton"] = nation_map["French"]
nation_map["Britain's"] = nation_map["British"]
nation_map["Britsih"] = nation_map["British"]
nation_map["California"] = nation_map["US"]
nation_map["Californian"] = nation_map["US"]
nation_map["Cantonese"] = nation_map["Chinese"]
nation_map["Caribbean"] = "Central America and the Caribbean"
nation_map["Catalan"] = nation_map["Spanish"]
nation_map["Caymanian"] = nation_map["Caribbean"]
nation_map["Ceylon"] = nation_map["Sri Lankan"]
nation_map["Ceylonese"] = nation_map["Sri Lankan"]
nation_map["Chagossian"] = "Indian Ocean"
nation_map["Chechen"] = nation_map["Russian"]
nation_map["Cherokee"] = nation_map["US"]
nation_map["Chilian"] = nation_map["Chilean"]
nation_map["China"] = nation_map["Chinese"]
nation_map["Chiricahua"] = nation_map["US"]
nation_map["Chuvash"] = nation_map["Russian"]
nation_map["Circassian"] = nation_map["Russian"]
nation_map["Columbian"] = nation_map["Colombian"]
nation_map["Congo"] = nation_map["Congolese"]
nation_map["Congoleze"] = nation_map["Congolese"]
nation_map["Cornish"] = nation_map["British"]
nation_map["Costan Rican"] = nation_map["Costa Rican"]
nation_map["Côte d'Ivoire"] = nation_map["Ivorian"]
nation_map["Crimean"] = nation_map["Russian"]
nation_map["Croat"] = nation_map["Croatian"]
nation_map["Curaçaoan"] = nation_map["Dutch"]
nation_map["Curaçaon"] = nation_map["Dutch"]
nation_map["Dagestani"] = nation_map["Russian"]
nation_map["Dahomey"] = "Africa"
nation_map["Dijiboutian"] = nation_map["Djiboutian"]
nation_map["Dolgan"] = nation_map["Russian"]
nation_map["Dominica"] = nation_map["Caribbean"]
nation_map["England"] = nation_map["British"]
nation_map["Englist"] = nation_map["British"]
nation_map["European"] = "Europe"
nation_map["Falkland Islands"] = "South America"
nation_map["Falkland islands"] = nation_map["Falkland Islands"]
nation_map["Falkland"] = nation_map["Falkland Islands"]
nation_map["Faroese"] = nation_map["Danish"]
nation_map["Filipina"] = nation_map["Filipino"]
nation_map["Filipo"] = nation_map["Filipino"]  # verified entry
nation_map["Fillipina"] = nation_map["Filipino"]
nation_map["Finish"] = nation_map["Finnish"]
nation_map["Flemish"] = nation_map["Belgian"]
nation_map["Franch"] = nation_map["French"]  # verified entry
nation_map["Franco"] = nation_map["French"]
nation_map["Frenck"] = nation_map["French"]  # verified entry
nation_map["Fujianese"] = nation_map["Chinese"]
nation_map[
    "Gaelic"
] = "Europe"  # refers to sport of Gaelic football, otherwise language
nation_map["Galician"] = nation_map["Spanish"]
nation_map["Galápagos"] = nation_map["Ecuadorian"]  # entry for tortoise
nation_map["Geman"] = nation_map["German"]  # verified entry
nation_map["Germen"] = nation_map["German"]  # verified entry
nation_map["Ghanese"] = "Africa"
nation_map["Greenlandic"] = nation_map["Danish"]
nation_map["Guadeloupean"] = nation_map["Caribbean"]
nation_map["Guamanian"] = "Oceania"
nation_map["Guernsey"] = nation_map["British"]
nation_map["Guernseyan"] = nation_map["British"]
nation_map["Hawaiian"] = nation_map["US"]
nation_map["Hindi"] = nation_map["Indian"]
nation_map["Hindu"] = nation_map["Indian"]
nation_map["Hollywood"] = nation_map["US"]
nation_map["Hong Kong"] = nation_map["Chinese"]
nation_map["Houston"] = nation_map["US"]
nation_map["Huaorani"] = nation_map["Ecuadorian"]
nation_map["I Kiribati"] = nation_map["IKiribati"]
nation_map["Indin"] = nation_map["Indian"]  # verified entry
nation_map["Indo"] = nation_map["Indian"]
nation_map["Ingush"] = nation_map["Russian"]
nation_map["Italo"] = nation_map["Italian"]
nation_map["Ivoirian"] = "Africa"
nation_map["Javanese"] = nation_map["Indonesian"]
nation_map["Jersey"] = nation_map["British"]
nation_map["Kabardin"] = nation_map["Russian"]
nation_map["Kashmiri"] = nation_map["Indian"]
nation_map["Korean"] = "Asia"
nation_map["Kosovan"] = "Europe"
nation_map["Kosovar"] = "Europe"
nation_map["Kosovo"] = "Europe"
nation_map["Kurdish"] = "Asia"
nation_map["Lesothan"] = "Africa"
nation_map["Los Angeles"] = nation_map["US"]
nation_map["Louisiana"] = nation_map["US"]
nation_map["MGerman"] = nation_map["German"]  # verified entry
nation_map["Macanese"] = nation_map["Chinese"]
nation_map["Malayalam"] = nation_map["Indian"]
nation_map["Malayali"] = nation_map["Indian"]
nation_map["Malayan"] = nation_map["Malaysian"]
nation_map["Manx"] = nation_map["British"]
nation_map["Mexian"] = nation_map["Mexican"]
nation_map["Mississippi"] = nation_map["US"]
nation_map["Monegasque"] = nation_map["Monacan"]
nation_map["Montserrat"] = nation_map["Caribbean"]
nation_map["Montserratian"] = nation_map["Caribbean"]
nation_map["Myanmar"] = nation_map["Burmese"]
nation_map["New York"] = nation_map["US"]
nation_map["Ngarrindjeri"] = nation_map["Australian"]
nation_map["Ni Vanuatu"] = "Oceania"
nation_map["Nigirean"] = nation_map["Nigerian"]
nation_map["Niuean"] = nation_map["NZ"]
nation_map["Northern Ire"] = nation_map["Northern Irish"]
nation_map["Northern Ireland"] = nation_map["Northern Irish"]
nation_map["Norther Irish"] = nation_map["Northern Irish"]
nation_map["North Irish"] = nation_map["Northern Irish"]
nation_map["North American"] = "North America"
nation_map["North Island"] = nation_map["NZ"]
nation_map["Northern Mariana Island"] = "Oceania"
nation_map["Northern Mariana Islander"] = "Oceania"
nation_map["Nubian"] = nation_map["Sudanese"]
nation_map["Ottoman"] = nation_map["Turkish"]
nation_map["Paraguan"] = nation_map["Paraguayan"]  # verified entry
nation_map["People's Republic of China"] = nation_map["Chinese"]
nation_map["Pitcairn"] = "Oceania"
nation_map["Poliosh"] = nation_map["Polish"]  # verified entry
nation_map["Polis"] = nation_map["Polish"]  # verified entry
nation_map["Prussian"] = nation_map["German"]
nation_map["Punjabi"] = nation_map["Indian"]
nation_map["Quebec"] = nation_map["Canadian"]
nation_map["Québécois"] = nation_map["Canadian"]
nation_map["Republic of China"] = nation_map["Chinese"]
nation_map["Rhodesian"] = "Africa"
nation_map["Roman"] = nation_map["Italian"]
nation_map["Réunionese"] = nation_map["French"]
nation_map["S African"] = nation_map["South African"]
nation_map["Saban"] = nation_map["Caribbean"]
nation_map["Saharawi"] = "Africa"
nation_map["Sahrawi"] = nation_map["Saharawi"]
nation_map["Saint Helena"] = "South Atlantic"
nation_map["Saint Vincent"] = nation_map["Caribbean"]
nation_map["Saint Martin"] = nation_map["Caribbean"]
nation_map["Saint Pierre and Miquelon"] = "North America"
nation_map["Salvadorean"] = nation_map["Salvadoran"]
nation_map["Sanmarinese"] = nation_map["Sammarinese"]
nation_map["Santomean"] = nation_map["São Toméan"]
nation_map["Seychellian"] = nation_map["Seychellois"]
nation_map["Sicilian"] = nation_map["Italian"]
nation_map["Sicillian"] = nation_map["Italian"]
nation_map["Sikkimese"] = nation_map["Indian"]
nation_map["Sorbian"] = nation_map["German"]
nation_map["South Afican"] = nation_map["South African"]
nation_map["South Ossetian"] = nation_map["Georgian"]
nation_map["Soviet"] = "United Socialist Soviet Republic"
nation_map["Sri lankan"] = nation_map["Sri Lankan"]
nation_map["St Lucian"] = nation_map["Caribbean"]
nation_map["St Kitts and Nevis"] = nation_map["Kittian and Nevisian"]
nation_map["Sumatran"] = nation_map["Indonesian"]
nation_map["Swish"] = nation_map["Swiss"]  # verified entry
nation_map["Tahitian"] = "South Pacific"
nation_map["Tamil"] = "Indian subcontinent"
nation_map["Tasmanian"] = nation_map["Australian"]
nation_map["Telugu"] = nation_map["Indian"]
nation_map["Texas"] = nation_map["US"]
nation_map["Florida"] = nation_map["US"]
nation_map["Tibetan"] = nation_map["Chinese"]
nation_map["Tirkish"] = nation_map["Turkish"]
nation_map["Tirkish"] = nation_map["Turkish"]
nation_map["Trinbagonian"] = nation_map["Trinidadian"]
nation_map["Trinidad"] = nation_map["Trinidadian"]
nation_map["Turks"] = nation_map["Turkish"]
nation_map["U S"] = nation_map["US"]
nation_map["UAE's"] = nation_map["Emirati"]
nation_map["United Kingdom"] = nation_map["British"]
nation_map["Upper Silesian"] = nation_map["Polish"]
nation_map["Uruguyan"] = nation_map["Uruguayan"]
nation_map["Uyghur"] = "Asia"
nation_map["Ni Vanuatu"] = nation_map["Vanuatuan"]
nation_map["Wallisian"] = "Oceania"
nation_map["Xhosa"] = nation_map["South African"]
nation_map["Yazidi"] = "Asia"
nation_map["Yellowstone"] = nation_map["US"]
nation_map["Yugoslav"] = nation_map["Serbian"]
nation_map["Yugoslavia"] = nation_map["Serbian"]
nation_map["Yugoslavian"] = nation_map["Serbian"]
nation_map["Zairean"] = nation_map["Congolese"]
nation_map["Zanzibari"] = nation_map["Tanzanian"]

<IPython.core.display.Javascript object>

#### Example of Checking Rows with Unique Value

In [22]:
# Cell for checking rows of unique values -- used repeatedly while hard-coding above
check_value = "Yellowstone "

check_list = []
for index in df[df["info_2"].notna()].index:
    item = df.loc[index, "info_2"]
    if item:
        if item.startswith(check_value):
            check_list.append(index)
df.loc[check_list, :]

day   name                                             info  \
58424   6  O-Six  , 6, Yellowstone National Park gray wolf, shot.   

                                      link num_references  year     month  \
58424  https://en.wikipedia.org/wiki/O-Six             16  2012  December   

      info_parenth info_1                               info_2 info_3 info_4  \
58424         None   None  Yellowstone National Park gray wolf   shot   None   

      info_5 info_6 info_7 info_8 info_9 info_10 info_11  age cause_of_death  \
58424   None   None   None   None   None    None    None  6.0           None   

      place_1 place_2  
58424     NaN     NaN

<IPython.core.display.Javascript object>

#### Appending Other Species List for New Species Observed During Hard-coding

In [23]:
# Adding more species to other_species
other_species_df = pd.read_csv("other_species.csv")
other_species = other_species_df["species"].tolist()
other_species.append("kiwi")
other_species.append("stallion")
other_species.append("colt")
other_species.append("Thoroughbred hurdler")
other_species.append("race horse")
other_species.append("racing filly")
other_species.append("wolf")

<IPython.core.display.Javascript object>

#### Observations:
- That was more brute force than we would like to be using, but as the nationality descriptions are mixed in with other features, the approach was chosen over using an approximate match approach such as the fuzzywuzzy library.
- We will proceed to re-run our code searching `info_2`.

#### Checking `info_2` for `place_1` with Updated `nation_map`

In [24]:
%%time

# Column to check
column = "info_2"

# Extract to column
extract_to = "place_1"

# Dataframe to check
dataframe = df[(df[extract_to].isna()) & (df[column].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if item.startswith(nationality):
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(nationality, "").strip()
            )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 7.31 s
Wall time: 7.32 s


<IPython.core.display.Javascript object>

In [25]:
# Check a sample of rows
df[df[extract_to].notna()].sample(2)

day               name                                         info  \
89092  21       Larry Wright                  , 77, American cartoonist .   
231     5  Joachim Halupczok  , 25, Polish racing cyclist, heart failure.   

                                                          link num_references  \
89092  https://en.wikipedia.org/wiki/Larry_Wright_(cartoonist)              6   
231            https://en.wikipedia.org/wiki/Joachim_Halupczok              3   

       year     month info_parenth info_1          info_2         info_3  \
89092  2017       May           ()   None      cartoonist           None   
231    1994  February         None   None  racing cyclist  heart failure   

      info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  \
89092   None   None   None   None   None   None    None    None  77.0   
231     None   None   None   None   None   None    None    None  25.0   

      cause_of_death                   place_1 place_2  
89092           None  United States of America     NaN  
231             None                    Poland     NaN

<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for `place_1`

In [26]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')

There are 643 remaining missing values for place_1.



<IPython.core.display.Javascript object>

#### Obervations:
- We have narrowed down the missing values for `place_1` values to just under 650.
- Let us take another look at the remaining unique values for `info_2`.

#### Checking Remaining Unique Values for `info_2` Where `place_1` Value is Missing

In [27]:
# Checking remaining unique values for info_2 where place_1 value is missing
df[df["place_1"].isna()]["info_2"].unique()

array(['Royal Netherlands Navy vice admiral', 'President of Laos',
       'Governor general of the Bahamas',
       'Amateur violinist and philanthropist',
       'West German long distance runner and Olympian',
       'Composer and music editor',
       'President of the Yemen Arab Republic',
       'Native American tribal leader', 'Prime Minister of Rwanda',
       'Male Hungarian international table tennis player',
       '37th President of the United States',
       'Queen of Jordan as the wife of King Talal',
       'Prime Minister of Zaire under Mobutu Sese Seko',
       'President of Burma and writer',
       'Founder and first leader of North Korea',
       'Moravian American classical pianist', 'President of Ciskei',
       'Prime Minister of Nepal', 'President of Palau',
       'Poet and an Esperantist professor', 'East German politician',
       'Jewish rabbi', '17th Indian Army officer and Chief of Staff',
       'All American baseball player',
       'pioneering American b

<IPython.core.display.Javascript object>

#### Observations:
- We can see some country names embedded in the middle of the values, so we will search there next, for nationality or country.
- In this iteration, we will allow the first value found to be overwritten and accept the second nationality value, in cases where there are two values.  Ultimately, the second value is the one we will maintain for all entries.

#### Checking `info_2` for `place_1` via Nationality or Country within Value

In [28]:
%%time

# Column to check
column = "info_2"

# Extract to column
extract_to = "place_1"

# Dataframe to check
dataframe = df[(df[extract_to].isna()) & (df[column].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if nationality in item or country in item:
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(nationality, "").strip().replace(country, '').strip()
            )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 2.16 s
Wall time: 2.17 s


<IPython.core.display.Javascript object>

In [29]:
# Check a sample of rows
df[df[extract_to].notna()].sample(2)

day            name  \
15789   10  Hafez al-Assad   
110750  11     Ruth Mandel   

                                                                                                              info  \
15789                                                                                    , 69, President of Syria.   
110750  , 81, Austrian-born American political scientist, women's advocate and Holocaust survivor, ovarian cancer.   

                                                link num_references  year  \
15789   https://en.wikipedia.org/wiki/Hafez_al-Assad            190  2000   
110750     https://en.wikipedia.org/wiki/Ruth_Mandel             12  2020   

        month info_parenth info_1               info_2  \
15789    June         None   None         President of   
110750  April         None   None  political scientist   

                                         info_3          info_4 info_5 info_6  \
15789                                      None            None   None   None   
110750  women's advocate and Holocaust survivor  ovarian cancer   None   None   

       info_7 info_8 info_9 info_10 info_11   age cause_of_death  \
15789    None   None   None    None    None  69.0           None   
110750   None   None   None    None    None  81.0           None   

                         place_1 place_2  
15789                      Syria     NaN  
110750  United States of America     NaN

<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for `place_1`

In [30]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')

There are 285 remaining missing values for place_1.



<IPython.core.display.Javascript object>

#### Observations:
- We have 285 remaining missing values for `place_1`.
- After checking these rows, we can update the dictionary again.

In [31]:
# Checking remaining unique values for info_2 where place_1 value is missing
df[df["place_1"].isna()]["info_2"].unique()

array(['Governor general of the Bahamas',
       'Amateur violinist and philanthropist',
       'Composer and music editor',
       'Prime Minister of Zaire under Mobutu Sese Seko',
       'President of Ciskei', 'Poet and an Esperantist professor',
       'Jewish rabbi',
       "native leader and historian of the Metepenagiag Mi'kmaq Nation",
       'sidecarcross rider and the first ever Sidecarcross World Championship',
       'Major League Baseball player', 'professional wrestler',
       'Field hockey player', 'fifth dean of the Harvard Business School',
       'Racecar driver', 'longtime chairman of Liverpool F C',
       'Orthodox Jewish rabbi',
       'Royal Air Force Air marshal and flying ace during World War II',
       'Archdeacon of Halifax',
       'inspector general and executive director of the CIA',
       'broadcast journalist for NBC News who was news anchor of',
       'younger son of Ayatollah Ruhollah Khomeini and father of Hassan Khomeini',
       'actor and occult

<IPython.core.display.Javascript object>

#### Hard-coding Additional Variations on Nationality

In [32]:
# Hard-coding remaining unique nationality/location descriptors
nation_map["Bahamas"] = nation_map["Bahamian"]
nation_map["Zaire"] = nation_map["Zairean"]
nation_map["Ciskei"] = "Africa"
nation_map["Metepenagiag Mi'kmaq"] = nation_map["Canadian"]
nation_map["Major League"] = nation_map["US"]
nation_map["Liverpool"] = nation_map["British"]
nation_map["Halifax"] = nation_map["Canadian"]
nation_map["CIA"] = nation_map["US"]
nation_map["Major Leagues"] = nation_map["US"]
nation_map["Levi Strauss & Co"] = nation_map["US"]
nation_map["Miami"] = nation_map["US"]
nation_map["Heaven's Gate"] = nation_map["US"]
nation_map["Royal Navy"] = nation_map["British"]
nation_map["Luftwaffe"] = nation_map["German"]
nation_map["IRA"] = nation_map["Irish"]
nation_map["Ashanti"] = "Africa"
nation_map["White House"] = nation_map["US"]
nation_map["Boeing"] = nation_map["US"]
nation_map["Red Army"] = nation_map["Soviet"]
nation_map["Yokohama"] = nation_map["Japanese"]
nation_map["Tasmania"] = nation_map["Tasmanian"]
nation_map["Harvard"] = nation_map["US"]
nation_map["Malaya"] = nation_map["Malayan"]
nation_map["Pennsylvania"] = nation_map["US"]
nation_map["House of Saud"] = nation_map["Saudi"]
nation_map["Faroe Islands"] = nation_map["Faroese"]
nation_map["Kentucky"] = nation_map["US"]
nation_map["Norfolk"] = nation_map["British"]  # verified entry
nation_map["Rhodesia"] = nation_map["Rhodesian"]
nation_map["Royal Marines"] = nation_map["British"]
nation_map["Rhode Island"] = nation_map["US"]
nation_map["Jerusalem"] = nation_map["Israeli"]
nation_map["Kriegsmarine"] = nation_map["German"]
nation_map["Nazi"] = nation_map["German"]
nation_map["South Carolina"] = nation_map["US"]
nation_map["Rashtriya Swayamsevak Sangh"] = nation_map["Indian"]
nation_map["Ontario"] = nation_map["Canadian"]
nation_map["London"] = nation_map["British"]
nation_map["McDonalds"] = nation_map["US"]
nation_map["Kleiner"] = nation_map["US"]
nation_map["Worldcom"] = nation_map["US"]
nation_map["Reino Aventura"] = nation_map["Mexican"]
nation_map["Palm Beach"] = nation_map["US"]
nation_map["Lawrence Radiation Laboratory"] = nation_map["US"]
nation_map["Royal Air Force"] = nation_map["British"]
nation_map["Britain"] = nation_map["British"]
nation_map["american"] = nation_map["US"]
nation_map["St  Lucian"] = nation_map["St Lucian"]
nation_map["Cook Islands"] = "Oceania"
nation_map["E O  Green Junior High School"] = nation_map["US"]
nation_map["Cook Islander"] = "Oceania"
nation_map["western lowland"] = "Africa"
nation_map["Saharan"] = "Africa"
nation_map["St  Kitts and Nevis"] = nation_map["Kittian"]
nation_map["Jammu and Kashmir"] = "Indian subcontinent"
nation_map["Cook Island"] = "Oceania"
nation_map["New Caledonian"] = "Oceania"

<IPython.core.display.Javascript object>

#### Adding More Values to `other_species`

In [33]:
# Adding more species to other_species
other_species.append("gorilla")
other_species.append("orca")
other_species.append("mammal")
other_species.append("triple crown winner")

<IPython.core.display.Javascript object>

#### Re-checking `info_2` for `place_1` via Nationality or Country within Value after Updating `nation_map`

In [34]:
%%time

# Column to check
column = "info_2"

# Extract to column
extract_to = "place_1"

# Dataframe to check
dataframe = df[(df[extract_to].isna()) & (df[column].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if nationality in item or country in item:
            df.loc[index, extract_to] = country
            df.loc[index, column] = (
                df.loc[index, column].replace(nationality, "").strip().replace(country, '').strip()
            )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 1.02 s
Wall time: 1.02 s


<IPython.core.display.Javascript object>

In [35]:
# Check a sample of rows
df[df[extract_to].notna()].sample(2)

day                name  \
121758  25  Zhenia Vasylkivska   
59968   17      Manoranjan Das   

                                                         info  \
121758  , 92, Ukrainian poet, translator and literary critic.   
59968                                , 89, Indian playwright.   

                                                    link num_references  year  \
121758  https://en.wikipedia.org/wiki/Zhenia_Vasylkivska              6  2021   
59968       https://en.wikipedia.org/wiki/Manoranjan_Das              8  2013   

           month info_parenth info_1      info_2  \
121758     April         None   None        poet   
59968   February         None   None  playwright   

                                info_3 info_4 info_5 info_6 info_7 info_8  \
121758  translator and literary critic   None   None   None   None   None   
59968                             None   None   None   None   None   None   

       info_9 info_10 info_11   age cause_of_death  place_1 place_2  
121758   None    None    None  92.0           None  Ukraine     NaN  
59968    None    None    None  89.0           None    India     NaN

<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for `place_1`

In [36]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')

There are 201 remaining missing values for place_1.



<IPython.core.display.Javascript object>

#### Observations:
- There are ~200 remaining missing values for `place_1`, which we do not expect to find in `info_2`.
- We can now move on to searching for `place_2` in `info_2`, again allowing here the first extracted value of `place_2` to be overwritten, if necessary.

#### Checking `info_2` for `place_2` via Nationality or Country Name within Value

In [37]:
%%time

# Column to check
column = "info_2"

# Extract to column
extract_to = "place_2"

# Dataframe to check
dataframe = df[
    (df["place_1"].notna()) & (df[extract_to].isna()) & (df[column].notna())]

# For loop to extract nation data to place column
for nationality, country in nation_map.items():
    for index in dataframe.index:
        item = df.loc[index, column]
        if nationality in item or country in item:
            df.loc[index, extract_to] = country
            df.loc[index, column].replace(nationality, "").strip().replace(country, '').strip()
            
# Chime notification when cell successfully executes
chime.success()

CPU times: total: 7min 2s
Wall time: 7min 2s


<IPython.core.display.Javascript object>

In [38]:
# Check a sample of rows
df[df[extract_to].notna()].sample(2)

day               name  \
17987  18  Seán Mac Stíofáin   
72861   2          Noel Cobb   

                                                                    info  \
17987         , 73, English-Irish chief of staff of the Provisional IRA.   
72861  , 76, American-born British philosopher, psychologist and author.   

                                                                 link  \
17987  https://en.wikipedia.org/wiki/Se%C3%A1n_Mac_St%C3%ADof%C3%A1in   
72861                         https://en.wikipedia.org/wiki/Noel_Cobb   

      num_references  year    month info_parenth info_1  \
17987             12  2001      May         None   None   
72861              5  2015  January         None   None   

                                            info_2                   info_3  \
17987  Irish chief of staff of the Provisional IRA                     None   
72861                          British philosopher  psychologist and author   

      info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  \
17987   None   None   None   None   None   None    None    None  73.0   
72861   None   None   None   None   None   None    None    None  76.0   

      cause_of_death                                               place_1  \
17987           None  United Kingdom of Great Britain and Northern Ireland   
72861           None                              United States of America   

                                                    place_2  
17987                                               Ireland  
72861  United Kingdom of Great Britain and Northern Ireland

<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for `place_1` and Number of Entries with Values for `place_2`

In [39]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')
print(f'There are {df["place_2"].notna().sum()} entries with values for place_2.')

There are 201 remaining missing values for place_1.

There are 6988 entries with values for place_2.


<IPython.core.display.Javascript object>

#### Observations:
- We are finished searching `info_2` for place values and can move on to searching the other numbered `info` columns and `info_parenth` for these values.

#### Checking `info_3` through `info_11` and `info_parenth` for Remaining `place_1` Values

In [41]:
%%time

# Columns to search
cols_lst = [
    'info_3',
    'info_4',
    'info_5',
    'info_6',
    'info_7',
    'info_8',
    'info_9',
    'info_10',
    'info_11',
    'info_parenth'
]

# Extract to column
extract_to = "place_1"

# For loop to extract nation data to place column
for column in cols_lst:
    dataframe = df[(df[extract_to].isna()) & (df[column].notna())]
    for nationality, country in nation_map.items():
        for index in dataframe.index:
            item = df.loc[index, column]
            if item.startswith(nationality):
                df.loc[index, extract_to] = country
                df.loc[index, column] = (
                    df.loc[index, column].replace(nationality, "").strip()
                )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 531 ms
Wall time: 516 ms


<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for `place_1`

In [43]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')

There are 178 remaining missing values for place_1.



<IPython.core.display.Javascript object>

#### Observations:
- That iteration captured close to 25 values.
- We will repeat it, but checking for values that start with the country value.

#### Re-checking `info_3` through `info_11` and `info_parenth` for `place_1` Using Country Value

In [44]:
%%time

# Extract to column
extract_to = "place_1"

# For loop to extract nation data to place column
for column in cols_lst:
    dataframe = df[(df[extract_to].isna()) & (df[column].notna())]
    for country in nation_map.values():
        for index in dataframe.index:
            item = df.loc[index, column]
            if item.startswith(country):
                df.loc[index, extract_to] = country
                df.loc[index, column] = (
                    df.loc[index, column].replace(country, "").strip()
                )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 453 ms
Wall time: 447 ms


<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for `place_1`

In [45]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')

There are 178 remaining missing values for place_1.



<IPython.core.display.Javascript object>

#### Observations:
- There were no matches for starting with the country value.
- Now, we will repeat searching for nation or country embedded in the value, allowing the first extracted value to be overwritten, where necessary.

#### Re-checking info_3 through info_11 and info_parenth for place_1 via Nationality or Country within Value

In [46]:
%%time

# Extract to column
extract_to = "place_1"

# For loop to extract nation data to place column
for column in cols_lst:
    dataframe = df[(df[extract_to].isna()) & (df[column].notna())]
    for nationality, country in nation_map.items():
        for index in dataframe.index:
            item = df.loc[index, column]
            if nationality in item or country in item:
                df.loc[index, extract_to] = country
                df.loc[index, column] = (
                    df.loc[index, column].replace(nationality, "").strip().replace(country, '').strip()
                )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 422 ms
Wall time: 418 ms


<IPython.core.display.Javascript object>

#### Checking Remaining Missing Values for `place_1`

In [50]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')

# Checking a sample of rows with missing place_1
df[df["place_1"].isna()].sample(5)

There are 171 remaining missing values for place_1.



day            name  \
22454  14           Dolly   
12037   3  Harry Jannotti   
2916   14     Frank Blair   
11699   2     Alam Channa   
94330  21     Kurt Hansen   

                                                                                info  \
22454  , 6, the world's first cloned mammal, euthanization following a lung disease.   
12037                                           , 74, Democratic politician, cancer.   
2916                , 79, broadcast journalist for NBC News who was news anchor of .   
11699                                                , 45, tall man, kidney failure.   
94330                                                            , 90, footballer[9]   

                                                                    link  \
22454                        https://en.wikipedia.org/wiki/Dolly_(sheep)   
12037                       https://en.wikipedia.org/wiki/Harry_Jannotti   
2916              https://en.wikipedia.org/wiki/Frank_Blair_(journalist)   
11699                          https://en.wikipedia.org/wiki/Alam_Channa   
94330  https://en.wikipedia.org/wiki/Kurt_Hansen_(footballer,_born_1928)   

      num_references  year      month info_parenth info_1  \
22454             48  2003   February         None   None   
12037              8  1998  September         None   None   
2916              11  1995      March         None   None   
11699              6  1998       July         None   None   
94330              2  2018   February         None   None   

                                                         info_2  \
22454                           the world's first cloned mammal   
12037                                     Democratic politician   
2916   broadcast journalist for NBC News who was news anchor of   
11699                                                  tall man   
94330                                             footballer[9]   

                                       info_3 info_4 info_5 info_6 info_7  \
22454  euthanization following a lung disease   None   None   None   None   
12037                                  cancer   None   None   None   None   
2916                                     None   None   None   None   None   
11699                          kidney failure   None   None   None   None   
94330                                    None   None   None   None   None   

      info_8 info_9 info_10 info_11   age cause_of_death place_1 place_2  
22454   None   None    None    None   6.0           None     NaN     NaN  
12037   None   None    None    None  74.0           None     NaN     NaN  
2916    None   None    None    None  79.0           None     NaN     NaN  
11699   None   None    None    None  45.0           None     NaN     NaN  
94330   None   None    None    None  90.0           None     NaN     NaN

<IPython.core.display.Javascript object>

#### Obervations:
- We have completed the search for `place_1` and the remaining values do appear to be lacking that data.
- The next step is to check these columns for any `place_2` values.
- As before, we will allow the first extracted value to be overwritten, if necessary.

#### Checking `info_3` through `info_11` and `info_parenth` for Remaining `place_2` Values

In [51]:
%%time

# Columns to search
cols_lst = [
    'info_3',
    'info_4',
    'info_5',
    'info_6',
    'info_7',
    'info_8',
    'info_9',
    'info_10',
    'info_11',
    'info_parenth'
]

# Extract to column
extract_to = "place_2"

# For loop to extract nation data to place column
for column in cols_lst:
    dataframe = df[(df['place_1'].notna()) & (df[extract_to].isna()) & (df[column].notna())]
    for nationality, country in nation_map.items():
        for index in dataframe.index:
            item = df.loc[index, column]
            if nationality in item or country in item:
                df.loc[index, extract_to] = country
                df.loc[index, column] = (
                    df.loc[index, column].replace(nationality, "").strip().replace(country, '').strip()
                )

# Chime notification when cell successfully executes
chime.success()

CPU times: total: 6min 49s
Wall time: 6min 49s


<IPython.core.display.Javascript object>

#### Final Counts of Missing Values for `place_1` and Number of Entries with Values for `place_2`

In [52]:
# Checking number of remaining missing values for place_1
print(f'There are {df["place_1"].isna().sum()} remaining missing values for place_1.\n')
print(f'There are {df["place_2"].notna().sum()} entries with values for place_2.')

There are 171 remaining missing values for place_1.

There are 15404 entries with values for place_2.


<IPython.core.display.Javascript object>

#### Observations:
- We are finished with extracting the place values.
- It's time to save our dataframe and start a new notebook before extracting `known_role` and `cause_of_death` values.
- Exporting `other_species` and `nation_map` is also a good idea, at this time.

## Exporting Dataset to SQLite Database [wp_life_expect_clean3.db]()

In [53]:
# Saving complete raw dataset in a SQLite database
conn = sql.connect("wp_life_expect_clean3.db")
df.to_sql("wp_life_expect_clean3", conn, index=False)

132652

<IPython.core.display.Javascript object>

## Saving nation_map to a Pickle File [nation_map.pkl]()

In [54]:
# Create a binary pickle file
f = open("nation_map.pkl", "wb")

# Write the dictionary to pickle file
pickle.dump(nation_map, f)

# close file
f.close()

<IPython.core.display.Javascript object>

## Exporting other_species List to [other_species.csv]()

In [55]:
# Convert other_species list to dataframe and save to csv -- overwriting existing csv file
other_species_df = pd.DataFrame({"species": other_species})
other_species_df.to_csv("other_species.csv", index=False)

<IPython.core.display.Javascript object>